provisioning/device/src/transport_state_machine.ts

```diff
@@ -0,0 +1,282 @@
+  }
+
+  /* istanbul ignore next */
+  _doOperationStatusQueryForFsm(registrationId: string, operationId: string, callback: (err?: Error, responseBody?: any, result?: any, pollingInterval?: number) => void): void {
+    throw new errors.NotImplementedError('_doOperationStatusQueryForFsm not implemented in this transport');
+  }
+
+  /* istanbul ignore next */
+  _getErrorFromResultForFsm(result: any): any {
+    throw new errors.NotImplementedError('_getErrorFromResultForFsm not implemented in this transport');
+  }
+
+
+  connect(callback: (err?: Error) => void): void {
+    debug('connect called');
+    this._fsm.handle('connect', callback);
+  }
+
+  register(registrationId: string, authorization: string | X509, requestBody: any, forceRegistration: boolean, callback: Provisioning.ResponseCallback): void {
+    debug('register called for registrationId "' + registrationId + '"');
+    this._fsm.handle('register', callback, registrationId, authorization, requestBody, forceRegistration);
+  }
+
+  disconnect(callback: (err: Error) => void): void {
+    debug('disconnect called');
+    this._fsm.handle('disconnect', callback);
+  }
+
+  private _cancelCurrentOperation(callback: (err?: Error) => void): void {
+    if (this._pollingTimer != null) {
+      debug('_cancelCurrentOperation: stopping polling timer');
+      clearTimeout(this._pollingTimer);
+      this._pollingTimer = null;
+    }
+
+    if (this._registrationInProgress()) {
+      debug('_cancelCurrentOperation: operation is in progress.  Cancelling.');
+      let _callback = this._registrationCallback;
+      this._registrationCallback = null;
+      _callback(new errors.OperationCancelledError());
+    }
+
+    callback();
+  }
+  private _handleOperationStatusResponse(registrationId: string,  responseBody: any, result: any, pollingInterval: number, callback: any): void {
+    if (responseBody) {
+      debug('received response from service:' + JSON.stringify(responseBody));
+      switch (responseBody.status.toLowerCase()) {
+        case 'assigned': {
+          /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_017: [ If the registration request returns a successful, `register` shall emit an 'operationStatus' event passing the body of the response. ] */
+          this.emit('operationStatus', responseBody);
+          /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_014: [ If the registration request returns with status==assigned, it shall call `callback` with null, the response body, and the protocol-specific result. ] */
+          /* Codes_SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_020: [ If `_doOperationStatusQueryForFsm` succeeds with status==Assigned, `register` shall complete and pass the body of the response and the protocol-spefic result to the `callback`. ] */
+          this._fsm.transition('connected', callback, null, responseBody, result);
```

Comment:
>this._fsm.transition('connected' [](start = 10, length = 32)

calling transition from outside the state machine feels sketchy - you should be calling handle('something') and let the state machine handle it correctly depending on what state it's in.